packages/react-i18n/src/index.tsx

```diff
@@ -21,35 +32,25 @@ interface Props {
 	locale: string;
 
 	/**
-	 * A callback that resolves with the translations data
+	 * Locale data
 	 */
-	onLocaleChange?( newLocale: string | undefined ): Promise< object >;
+	localeData?: object;
 }
 export const I18nProvider: React.FunctionComponent< Props > = ( {
 	children,
 	locale,
-	onLocaleChange = () => Promise.resolve( {} ),
+	localeData = {},
 } ) => {
+	updateLocaleOnChange( locale, localeData );
```

Comment:
The render function should be pure, without side effects. But `updateLocaleOnChange` is updating a global variable.

And the `lastLocale` variable is shared by all components that use `react-i18n`. It creates another obstacle on the way to working SSR, because different React trees rendered in different locales in parallel (responding to parallel HTTP requests) will fight over this variable.

I think that the `useI18n` hook and context shouldn't call `setLocaleData()` at all. In Gutenboarding for example, let's make it part of the code that figures out the locale and fetches the translation file. `setLocaleData()` belongs there naturally, IMO:
```js
const locale = await figureOutUserLocale();
const localeData = await fetchTranslationFile( locale );
i18n.setLocaleData( localeData );

ReactDOM.render(
  <I18nProvider locale={ locale } />
    <App />
  </I18nProvider>
);
```

Reply:
For context: I'm expecting most of this to be thrown away when https://github.com/WordPress/gutenberg/issues/19210 is resolved and I don't really believe this package provides much value until it can support the features you've mentioned.

I'll make the changes you suggested.

Reply:
I agree: when we have multiple instances available, we'll simply create a new `i18n` instance for every locale, load the data into it, and pass it as the context value. Is that what you had in mind?

Reply:
Remove locale setting from 7a3eef9fe91370833eabed77a28b713ed784cf30 which makes the package very simple.

Reply:
> I agree: when we have multiple instances available, we'll simply create a new `i18n` instance for every locale, load the data into it, and pass it as the context value. Is that what you had in mind?

Yes, that's the kind of detail I was expecting to handle in this library. I was probably getting ahead of myself with the current functionality of `@wordpress/i18n` 😉